### PR TITLE
feat(menu): Submenu

### LINKED ISSUE #542: [RFC] Submenu functionality in Menu component
### Problem description
Menu should allow defining submenus. Each menu item can have a submenu.
Submenu should follow ARIA specification and best practices:
https://www.w3.org/TR/wai-aria-practices/#menu
https://www.w3.org/TR/wai-aria-practices/examples/menubar/menubar-1/menubar-1.html

especially with regards to the rendered DOM strucutre and keyboard navigation:
- Submenu should be rendered as `<ul>` inside of the MenuItem `wrapper` slot. Menu items of that submenu should render normally as `<li>` items.
- Submenu is always vertical
- Home and End keys should not be used as the menu implements circular navigation.

As opposed to the specification described in the links above, submenu should not open when the user hovers on the menu item with a submenu. It should open when the user clicks on such a menu item or if he uses keyboard navigation keys for subemnu open.

### API Proposal
- add `menu` prop to the `MenuItem` component. User can pass a `Menu` shorthand to it.

### MVP
There has already been two PRs implementing this. Both had good and bad sides and the good parts were merged into #539.

### PR DESCRIPTION
**Pull request by jurokapsiar**
Based on two previous menu/submenu PRs. Fixes https://github.com/stardust-ui/react/issues/542

Missing things:
- [x] document click / focus out should close submenu
- [x] action should close submenu
- [x] left/right key should move to the next/previous menu item according to ARIA practices
- [x] submenu indicator
- [x] commented out overflow, not sure why it was there, need to verify
- [x] verify wrapper render

Update by @mnajdova 
Some questions:
- I needed to add key to the wrapper element created, so in the defaultProps I added key that is generated by a new added dependency `uniqid`. Are we ok with this?
- The tests are failing, I guess because I added `Ref` wrapper around the `ElementType`. How can be tackle this? Is the wrapper maybe causing this?
- How are we handling event handlers when we have wrapper? There is issue with the onClick event, which is added on the wrapper, but on the anchor is added only if the wrapper is not added. Am I doing something wrong here?

The questions were resolved thanks to @layershifter

All accessibility requirements are added, let me know if you find some issue with it.

### CODE REVIEW

**Comment by layershifter** [comment]
Should we use the `kind` prop to simplify the render logic, #512?

File: src/themes/teams/components/Menu/menuStyles.ts
Comment (jurokapsiar) [inline src/themes/teams/components/Menu/menuStyles.ts]: this needs to be verified - it breaks submenu when oveflow is hidden

File: src/themes/teams/components/Menu/menuVariables.ts
Comment (jurokapsiar) [inline src/themes/teams/components/Menu/menuVariables.ts]: this should come from theme/site

File: src/components/Menu/MenuItem.tsx
Comment (sophieH29) [inline src/components/Menu/MenuItem.tsx]: Keys should be assigned to actions in behaviors. I think we can have one action called ```performSubmenuClick``` which will call ```handleSubmenuClicked``` on Enter or Space. Same way as we have on MenuItem ```performClick``` line 231.

So basically in ```menuItemBehavior``` you add to key action

```js
keyActions: {
   ...
   
   submenu: {
      performSubmenuClick: {
        keyCombinations: [{ keyCode: keyboardKey.Enter }, { keyCode: keyboardKey.Spacebar }],
      },
    },
   ...
}
```

In ```MenuItem``` comp, add action handler (line 231)

```js
actionHandlers = {
    ...
    performSubmenuClick: (event) => this.handleSubmenuClicked(event, null)

    ...
}
```

NOTE: I see that in your handlers you receive props as second param. In case of ```keydown``` event, client code is handled in```getKeyDownHandlers.ts```, by calling ```_.invoke(props, 'onKeyDown', event)```, so we can just pass NULL as 2nd param in action handler.

and then line 188 where you create submenu

```js
const maybeSubmenu =
       ...
        Menu.create(menu, {
            defaultProps: {
              accessibility: submenuBehavior, 
              
            },
            overrideProps: {
              onClick: this.handleSubmenuClicked,
             ...accessibility.keyHandlers.submenu   // adds onkeyDown handler which will call 'performSubmenuClick'
            },
          })
        : null
```

Please let me know what you think.

File: src/lib/accessibility/Behaviors/Menu/submenuBehavior.ts
Comment (sophieH29) [inline src/lib/accessibility/Behaviors/Menu/submenuBehavior.ts]: Don't see any reason for now to have ```submenuBehavior``` as it is the same as ```menuBehavior```

File: src/components/Menu/MenuItem.tsx
Comment (layershifter) [inline src/components/Menu/MenuItem.tsx]: Please use a `createRef()` API there: https://github.com/stardust-ui/react/pull/584#discussion_r241463373

Comment (mnajdova) [reply src/components/Menu/MenuItem.tsx]: Done, thanks!

File: src/themes/teams/components/Menu/menuStyles.ts
Comment (mnajdova) [reply src/themes/teams/components/Menu/menuStyles.ts]: This is the PR that added it: https://github.com/stardust-ui/react/pull/360 Will sync with @Bugaa92 for finding alternative solution for the cropped borders, I have new ideas.

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [inline src/components/Menu/MenuItem.tsx:255]: note, this approach (presumably, taken from the Popup) has one subtle problem: although it solves the issue of item being opened and closed within the same frame, it also opens possibility for the following behavior:

- MenuItem has opened submenu, this opened prop value doesn't change anyhow for the next JS frame.

///------ JS FRAME --------////
- render method was called (but `submenuOpen ` wasn't changed changed anyhow, it is still `true`)
- updateOutsideClickSubscription
  -  this.outsideClickSubscription.unsubscribe(), no subscription on outside click (!!)
  - setTimeout(() => subscription will be made on the next frame)

- ... some other logic ..
   - note, there is no click subscription at this moment, but it should be (!!)

////---------------------------////

Would suggest to leave a fix for a separate PR, for both components.

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [inline src/components/Menu/MenuItem.tsx]: this one is easy to be confused with `openSubmenu` which we have in the same file. Verb, probably, would help to better differentiate - or past tense could be used. Lets see whether this is consistent with the rest of the code (and naming used), but I would rather like to eliminate this confusion.

Potential suggestions:
- `isSubmenuOpen`
- `submenuOpened`

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [inline src/components/Menu/MenuItem.tsx]: won't this override the effect of `submenuOpen={true}` prop being provided by the client? Have taken a look on the logic of `AutoControlledComponent` (`getAutoControlledStateValue`) and it seems to be pretty much the case

File: docs/src/examples/components/Menu/Types/MenuExampleWithSubmenu.shorthand.tsx
Comment (kuzhelov) [inline docs/src/examples/components/Menu/Types/MenuExampleWithSubmenu.shorthand.tsx:33]: 👍 for naming consistency from level to level in the nested tree (no props like `submenu`, etc)

File: src/components/Menu/MenuItem.tsx
Comment (mnajdova) [reply src/components/Menu/MenuItem.tsx]: Yep, we should never have initial value for e state that is part of the Autocontrolled prop. This was leftover from the prev PRs merged here. Will remove it. Thanks for the catch!

File: src/components/Menu/MenuItem.tsx
Comment (mnajdova) [reply src/components/Menu/MenuItem.tsx]: We never use `is` as a prefix for the boolean props, and the similar auto-controlled prop in the Popup, has the name `open`. But, I agree with your concerns, would rather rename it to `submenuOpened`, but not sure as it won't be consistent. What do you think?

File: src/components/Menu/MenuItem.tsx
Comment (mnajdova) [reply src/components/Menu/MenuItem.tsx:255]: Will create issue for this, after the PR will be merged

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [inline src/components/Menu/MenuItem.tsx:234]: sorry, just to understand what current limitations are. What will happen if `wrapper` will be evaluated to falsy value?
There won't be any submenu functionality supported in this case, right?

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [reply src/components/Menu/MenuItem.tsx]: lets rely on consistency first - by that we will ensure that problem will be fixed for all components and in the same way. If this name is consistent with other components' prop names, lets leave it 👍

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [inline src/components/Menu/MenuItem.tsx:189]: are we really sure that we would like to override client's `onClick` logic in that case?

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [inline src/components/Menu/MenuItem.tsx:205]: is there any possibility that it could be horizontal one? I mean, I might imagine it (horizontal menu and horizontal submenu) - this was the case for Word's toolbar, for example

File: src/components/Menu/MenuItem.tsx
Comment (mnajdova) [reply src/components/Menu/MenuItem.tsx:234]: Yes, because the submenu as an element is part of the wrapper (it cannot be part of the anchor tag, which is the element created if the wrapper is not provided). I agree that this is not obvious, maybe we should add that in the doc somewhere.

File: src/components/Menu/MenuItem.tsx
Comment (mnajdova) [reply src/components/Menu/MenuItem.tsx:189]: Actually, we are not overriding. The implementation of `handleClick` is the following:
```
  private handleClick = e => {
    this.performClick(e)
    _.invoke(this.props, 'onClick', e, this.props)
  }
```

File: src/components/Menu/MenuItem.tsx
Comment (mnajdova) [reply src/components/Menu/MenuItem.tsx:205]: According to the spec: https://github.com/stardust-ui/react/issues/542 the submenus should always be vertical. @jurokapsiar opinion on this?

File: docs/src/examples/components/Menu/Types/index.tsx
Comment (sophieH29) [inline docs/src/examples/components/Menu/Types/index.tsx]: I would name it ```Vertical menu with submenus```. Because both of the examples are vertical submenus :)

File: src/components/Menu/MenuItem.tsx
Comment (sophieH29) [inline src/components/Menu/MenuItem.tsx]: what is a use case to have an additional check on ```Esc``` button? We should avoid using keys in components, if you need a specific thing to happen on Esc, better create a separate action handler for that.

File: src/lib/accessibility/Behaviors/Menu/menuBehavior.ts
Comment (sophieH29) [inline src/lib/accessibility/Behaviors/Menu/menuBehavior.ts]: probably, we don't need this comment here :)

File: src/components/Menu/MenuItem.tsx
Comment (layershifter) [inline src/components/Menu/MenuItem.tsx]: We have a `doesNodeContainClick()` util.

File: src/components/Menu/MenuItem.tsx
Comment (mnajdova) [reply src/components/Menu/MenuItem.tsx]: Added new handler, thanks for the suggestion!

File: src/lib/accessibility/Behaviors/Menu/menuBehavior.ts
Comment (mnajdova) [reply src/lib/accessibility/Behaviors/Menu/menuBehavior.ts]: I left it intentionally for your guys, as I don't think we had final decision whether we wanted to have this in the other behaviors.

File: src/components/Menu/MenuItem.tsx
Comment (mnajdova) [reply src/components/Menu/MenuItem.tsx]: Changed, thanks!

File: docs/src/examples/components/Menu/Types/MenuExampleWithSubmenu.shorthand.tsx
Comment (kuzhelov) [inline docs/src/examples/components/Menu/Types/MenuExampleWithSubmenu.shorthand.tsx]: shouldn't it be part of the default styles? like

```
// menuItemStyles
'& .ui-menu {
  z-index: ..
}'
```

File: src/components/Menu/Menu.tsx
Comment (kuzhelov) [inline src/components/Menu/Menu.tsx:99]: here is another clear example - lets just take a look on the following code:

```
<Menu submenu={...} />
```

given this code, what we might expect to be passed as a `submenu`'s value? actually, I would expect items collection. This is what we have for `MenuItem`'s `menu`, but here - this is a boolean flag :)

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [inline src/components/Menu/MenuItem.tsx]: don't like this names duality - on the one side we have `menu`, on the other one `submenuOpen` - while, in fact, the same exact menu is implied by both. Consider the following expression as an example:

```if (menu && submenuOpen) { ... }```

 Suggest to rename to `menuOpen`

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [inline src/components/Menu/MenuItem.tsx]: how focus is passed in case if it is not a top-level menu? Are we implicitly relying on event bubbling here?

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [inline src/components/Menu/MenuItem.tsx]: `index` is only used for making this item `active` by this callback. We can remove `index` altogether and rename this prop to something that is more specific (and better reflects intent of this method).

Actually, the thing we have here is event that is fired in controlled mode of `MenuItem` - and, thus, we can name it as `onActiveChanged` and call when `MenuItem` would like to 'activate' itself

**Review by kuzhelov — COMMENTED** [review]
lets discuss mentioned points and try to simplify the approach a bit. 

Also, please, note that currently we have the following problem related to visual representation of focused item:

image

Could you say which one os focused now? :)

**Comment by mnajdova** [comment]
@kuzhelov thanks for the review. Sure, let's discuss the points, and for the focus styling (it is something that has been already implemented. Not sure what are the spec for this, but we can change the color of course, and update it later if that is not the correct one).

File: src/components/Menu/MenuItem.tsx
Comment (mnajdova) [reply src/components/Menu/MenuItem.tsx]: Fow now yes, let's consider implementing this with callback mechanism in a follow up PR.

File: src/components/Menu/Menu.tsx
Comment (mnajdova) [reply src/components/Menu/Menu.tsx:99]: Totally agree, for consistency I will leave it as this, and will create issue for renaming all boolean props to have the `is` prefix. Thanks for pointing this out.

File: docs/src/examples/components/Menu/Types/MenuExampleWithSubmenu.shorthand.tsx
Comment (mnajdova) [reply docs/src/examples/components/Menu/Types/MenuExampleWithSubmenu.shorthand.tsx]: I thought this was added only because of the docs experience, but it turns out that it is common practice to add z-index on the submenus. Will add this styles in the MenuItem, thanks!

File: docs/src/examples/components/Menu/Types/index.tsx
Comment (layershifter) [inline docs/src/examples/components/Menu/Types/index.tsx]: This example should be present, but we can leverage the attention to it there.

```diff
<ComponentExample
-  title="Vertical menu with submenus"
-  description="A vertical menu can have submenu."
  examplePath="components/Menu/Types/MenuExampleVerticalWithSubmenu"
/>
```

<details>
  <summary>Preview</summary>

image
</details>

File: src/components/Menu/Menu.tsx
Comment (layershifter) [inline src/components/Menu/Menu.tsx:69]: Should we export this interface?

File: src/components/Menu/Menu.tsx
Comment (layershifter) [inline src/components/Menu/Menu.tsx]: This condition is not valid, if we `activeIndex` is `string` it will be converted to `number`, in other cases it will be boolean:

```js
let activeIndex = 1
let index = 1
let active

active = typeof activeIndex === 'string' ? parseInt(activeIndex, 10) : activeIndex === index
console.log(typeof active) // boolean

activeIndex = '1'
active = typeof activeIndex === 'string' ? parseInt(activeIndex, 10) : activeIndex === index
console.log(typeof active) // number
```

File: src/components/Menu/Menu.tsx
Comment (kuzhelov) [reply src/components/Menu/Menu.tsx:69]: while understand the reason for this sentiment, it turns out that, actually, we should - as in the style functions we need to provide both state and props types, as one of the args of this style function is the result of `props` being merged to `state`:

```js
// this is the code that will be written by client
const buttonStyles = {
  // here props are of type (ButtonState & ButtonProps)
  root: ({ props, ... }) => { ... }
```

File: src/components/Menu/Menu.tsx
Comment (mnajdova) [reply src/components/Menu/Menu.tsx]: Yep, I am missing brackets
` typeof activeIndex === 'string' ? parseInt(activeIndex, 10) : activeIndex === index` => ` (typeof activeIndex === 'string' ? parseInt(activeIndex, 10) : activeIndex) === index
`

File: src/components/Menu/Menu.tsx
Comment (mnajdova) [reply src/components/Menu/Menu.tsx:69]: Agreed, will export it.

File: src/components/Menu/Menu.tsx
Comment (mnajdova) [reply src/components/Menu/Menu.tsx]: The active prop should always be boolean. Fixed.

**Comment by mnajdova** [comment]
All comments are resolved. I changed the look of the hovered/focused vs active styles. Couple of examples of the variants:
image
image
image
image
image

File: src/components/Menu/Menu.tsx
Comment (mnajdova) [reply src/components/Menu/Menu.tsx:69]: Exported

File: src/lib/accessibility/Behaviors/Menu/menuBehavior.ts
Comment (mnajdova) [reply src/lib/accessibility/Behaviors/Menu/menuBehavior.ts]: @sophieH29 @jurokapsiar should we add the direction prop to the toolbarBehavior, tabListBehavior as well, cus we have circular navigation there as well?

Comment (sophieH29) [reply src/lib/accessibility/Behaviors/Menu/menuBehavior.ts]: yes, we probably will add it in the future when needed. But don't see any reason to have this comment here.

File: docs/src/examples/components/Menu/Variations/MenuExampleIconOnlyPrimaryInverted.shorthand.tsx
Comment (kuzhelov) [inline docs/src/examples/components/Menu/Variations/MenuExampleIconOnlyPrimaryInverted.shorthand.tsx:17]: 👍

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [inline src/components/Menu/MenuItem.tsx]: nit: we can make this check (`this.state.menuOpen`) as the first step, because (`doesNodeContainClick`) are less efficient ones:

```js
if (!this.state.menuOpen) {
  return 
}
...
```

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [inline src/components/Menu/MenuItem.tsx:275]: thanks for comments, very helpful ones!

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [inline src/components/Menu/MenuItem.tsx]: for some cases we have this focus logic being run in the callback of the state change (which is correct), somewhere not. Lets make approach to that being consistent (or, if there is any reason why it is not, please, just let me know)

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [inline src/components/Menu/MenuItem.tsx]: it is hard to infer the difference of logic provided by this method vs subsequent `closeMenuAndFocusNextParentItem` one by using their names only - more than that, quite a lot of implementation code is similar. Maybe we could refactor it to be a single method with some parameter that will drive (and describe) this difference?

Comment (mnajdova) [reply src/components/Menu/MenuItem.tsx]: It was initially one method, but we have different handlings if the esc key was used, or right arrow, so that suggestions from behaviors point of view was that we should have different methods for these, as we don't want to use the check for the keys inside the component's logic.

File: src/components/Menu/MenuItem.tsx
Comment (mnajdova) [reply src/components/Menu/MenuItem.tsx]: Initially I had here check for which key was pressed, and I could not use the event in async manner there. That is not the case anymore, as we decided to use different handlers for different keys. We can move this safely in the callback now.

File: src/components/Menu/MenuItem.tsx
Comment (kuzhelov) [reply src/components/Menu/MenuItem.tsx:205]: as a proof for my point:

image

**Review by kuzhelov — APPROVED** [review]

File: src/components/Menu/MenuItem.tsx
Comment (mnajdova) [reply src/components/Menu/MenuItem.tsx:205]: Will be added in a follow up PR.

**Pull request merged**
